### PR TITLE
Implementing pdiff (mod persistence)

### PR DESCRIPTION
**Pull request by ASpoonPlaysGames**
Basic stuff to do:
- [x] Reading basic mod persistence and sending it to servers on client connect
- [x] Writing basic player mod persistence data to masterserver

More advanced (and painful) stuff to do:
- [x] Reading advanced mod persistence (adding to arrays, etc.)
- [x] Writing advanced mod persistence

Edit: must be merged alongside [NorthstarLauncher/148](https://github.com/R2Northstar/NorthstarLauncher/pull/148)

### CODE REVIEW

**Comment by ASpoonPlaysGames** [comment]
Current progress: reading and writing basic pdiff is functional, however I am yet to split the pdiff from the rest of the pdata in a way that conserves the remaining pdata, which could then be written to the pdata baseline. 

TLDR: you can do pdiff, but if you are doing pdiff you cant do normal pdata atm

**Comment by ASpoonPlaysGames** [comment]
pdiff now seems to be functional, for pdiffs that do not add to enums and stuff like that, idk where its going wrong in that regard

basically pdiffs like this are fine
`$PROP_START
$ENUM_START pDiffTest
	tdm
	cp
	at
	ctf
	lts
	ps
	ffa
	coliseum
	aitdm
	speedball
	mfd
	ttdm
	fra
	fd
$ENUM_END
int isPDiffWorking
string{9} testString
string{9} testString2`

But pdiffs like this break stuff:
`$ENUM_ADD gameModes
	fuck
$ENUM_END

$PROP_START
$ENUM_START pDiffTest
	tdm
	cp
	at
	ctf
	lts
	ps
	ffa
	coliseum
	aitdm
	speedball
	mfd
	ttdm
	fra
	fd
$ENUM_END
int isPDiffWorking
string{9} testString
string{9} testString2`

**Comment by ASpoonPlaysGames** [comment]
So basically, this *could* be considered ready for review? We would then release it in a sort of beta state?

**Comment by ASpoonPlaysGames** [comment]
Oh one more thing to note, when I create a string the game doesnt seem to like to write to the last char, always leaves it blank, idk if its like that with normal pdef or not though

**Comment by ASpoonPlaysGames** [comment]
This is now ready for testing I think, it does require the linked northstar launcher PR though

**Comment by ASpoonPlaysGames** [comment]
Yeah so there isn't really a way to limit pdiff size per player/mod, because of how pdiffs interact with each other and vanilla pdata, the size of the data stored in the MS is variable.

for example: if i set my primary weapon in my loadout to mp_weapon_peacekraber, it would get saved in the pdiff of the peacekraber's mod, however when i change it back to what it was before, it is no longer saved there, and therefore the pdiff data shrinks

**Comment by GeckoEidechse** [comment]
> I could enforce a pdiff size limit but tbh

Yeah we would need to have proper error handling where a too large pdiff would send an error message/code back to client that then parses it and shows the player the appropriate error :I

**Comment by ASpoonPlaysGames** [comment]
> a too large pdiff would send an error message/code back to client

annoyingly i dont think a client would ever actually get that message, since pdata is saved when the map changes or the player leaves, and it is also only ever networked to clients through the server, we never send pdata to clients directly

**Comment by ASpoonPlaysGames** [comment]
I might change the way that we get a hash for the pdiffs, as the current method results in "lost" data in the DB if the pdiff file ever changes, so if we ever use pdiff in things like Northstar.Custom and update the pdiff file it would be less than ideal. Probably going to look into making sure that the pdata doesn't get malformed when loading pdiffs after a pdiff file has been modified, and then probably hashing the mod name instead of the pdiff file

File: shared/modPersistentData.js [hunk 1]
```diff
@@ -0,0 +1,1047 @@
...
+		var pdefCopy = JSON.parse( JSON.stringify( DEFAULT_PDEF_OBJECT ) )
+
+		for ( let pdiffstr of pdiffs )
+		{
+			let pdiff
+			if ( pdiffstr )
+			{
+				try
+				{
+					let pdiffHash = crypto.createHash( "sha1" ).update( pdiffstr ).digest( "hex" )
```
Comment (ASpoonPlaysGames) [inline shared/modPersistentData.js:296]: Thinking of changing this to hash the mod name, because if a mod updates and adds anything to their pdiff file, all of the data for the users of the mod will be lost and waste space in the DB unnecessarily

File: persistent_player_data_version_231.pdef [hunk 2]
```diff
@@ -1469,7 +1469,10 @@ $STRUCT_START eFDPostGamePlayer
 	string{22}	xuid
 	int 		awardId
 	float 		awardValue
-	int 		suitIndex
+	// this used to be "int suitIndex" but that might cause issues with pdiffs that add more suits
+	pilotSuit 	suitIndex
+	// this is just here to avoid changing the size of the pdef without just appending data
+	bool 		unused[3]
```
Comment (ASpoonPlaysGames) [inline persistent_player_data_version_231.pdef:1475]: I'm not actually 100% sure if it should be this way around or the unused stuff first?

**Comment by GeckoEidechse** [comment]
This is still WIP, right?

**Comment by ASpoonPlaysGames** [comment]
> This is still WIP, right?

Unfortunately, yeah, Respawn added an arbitrary limit that kinda ruins pdiff, so it's probably going to need a decent size rework

**Comment by pg9182** [comment]
Closing for now as discussed on Discord.
